### ISSUE TITLE
Inserting a tensor into a python dict causes strange behavior

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#7733]
Reported from https://discuss.pytorch.org/t/why-tensor-is-hashable/18225 
```
T = torch.randn(5,5)
c = hash(T)  # i.e. c = 140676925984200
dic = dict()
dic[T] = 100
dic[c] 
RuntimeError: bool value of Tensor with more than one value is ambiguous.
```

I'm not sure if we support using Tensors as keys in python dicts; but Tensor does have a __hash__ method.

**Comment by zou3519** [comment]
We should disallow this and/or add a harder error message

**Comment by apaszke** [comment]
Well it all boils down to the implementation of `__bool__`, and this is the best message we can raise in general. Remember that dicts are basically hash tables, and hash tables need the keys to be hashable AND comparable. While you certainly can do `==` on tensors, it gives you a byte tensor, which will get its `__bool__` called, and this causes the code to fail (and it should, because it's unclear if you want `.all()` or `.any()` in general when you're doing it).

Getting rid of `__hash__` seems like a good idea.

**Comment by apaszke** [comment]
BTW what happens if you do that on a numpy array?

**Comment by mcarilli** [comment]
I noticed this in passing because it was next to an issue I filed.  Pytorch optimizers do use tensors as keys of the state dict:  
https://github.com/pytorch/pytorch/blob/master/torch/optim/optimizer.py#L134-L140
https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L95
so it appears to be useful, unless I'm misinterpreting something.

**Comment by soumith** [comment]
yea, @mcarilli is right, we probably have to keep `__hash__` around :-/

**Comment by apaszke** [comment]
Current `__hash__` is equivalent to using `id(tensor)` as a key, so there's really no reason to keep it. We can adjust optimizers to use `id` explicitly.

**Comment by weiyangfb** [comment]
So what is the decision now? To remove __hash__ and refactor optimizers as well as other files?

**Comment by apaszke** [comment]
Sounds good to me.

**Comment by rohan-varma** [comment]
I'm interested in contributing to Pytorch, and if this isn't being worked on I'd be happy to take a stab at it!

**Comment by weiyangfb** [comment]
@rohan-varma Go for it! Let us know if you need any guidance :)

**Comment by weiyangfb** [comment]
@rohan-varma Any updates so far? Do you need some helps?

**Comment by vishwakftw** [comment]
I can take this up, if it is free.

**Comment by weiyangfb** [comment]
@vishwakftw Yes, I haven't look into this much. Please feel free to take it up :)

**Comment by vishwakftw** [comment]
I don't understand the behavior of `hash()`. For an immutable type e.g., string:
```python
a = 'torch'
b = hash(a)
c = hash(hash(b)) # c != b (stated otherwise in the forums)
test = dict()
test[a] = 1
test[b] # KeyError
```
What should be the expected behavior? Although, since `torch.Tensor` is mutable, `__hash__` shouldn't exist.

**Comment by weiyangfb** [comment]
If you are worring about the behavior of `hash(Tensor)` and `__hash__()`. I think they should have consistent behaviors as mutable types. So we can probably remove `__hash__()` or raise an error whenever it gets called.

```
hash([1, 2]) # TypeError: unhashable type: 'list'

a = [1, 2]
print(a.__hash__()) # TypeError: 'NoneType' object is not callable
```

**Comment by mcarilli** [comment]
DistributedDataParallel and replicate.py (used by both DistributedDataParallel and DataParallel) also use tensors as dict keys:
https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L188
https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/replicate.py#L11
These are just examples I've stumbled across.  There may be any number of other instances where tensors were used as keys simply because it was a natural thing to do.  We do it quite a bit in Apex as well.  I don't mind adjusting our code to use id() explicitly, in response to whatever you guys decide, just giving a heads up that there may be a bit of pain initially as various usages of tensors as dict keys are tracked down.

**Comment by vishwakftw** [comment]
I also did notice that some of the tests also use tensors as a hashable type, while trying to make this change, but haven't found all of them yet.

**Comment by zou3519** [comment]
One way to find use cases is to throw an error in __hash__() and then run the pytorch tests, examples, and tutorials. This isn't comprehensive but it should be a good start.

**Comment by weiyangfb** [comment]
@vishwakftw Are you working on this? Or you wanna give it a try?

**Comment by vishwakftw** [comment]
I was working on this earlier, but I had some other work to do in the meanwhile. I can work on from next week onwards, once my internship's over.

**Comment by matt-gardner** [comment]
We ran into a bug in our code that was indirectly caused by this issue:

```python
>>> tensor = torch.Tensor([1,1])
>>> d = {(1,1): 1}
# pytorch 0.3:
>>> tuple(tensor) == (1, 1)
True
>>> tuple(tensor) in d
True
# pytorch 0.4:
>>> tuple(tensor) == (1,1)
True
>>> tuple(tensor) in d
False
```

With the addition of scalar tensors `tuple(tensor)` no longer returns a tuple of ints, it now returns a tuple of tensors.  These can be coerced to equal `(1, 1)`, but can't be used as lookups to a dictionary containing tuples of ints.  This breaks python's contract for [`__hash__` methods](https://docs.python.org/3/reference/datamodel.html#object.__hash__):

```
The only required property is that objects which compare equal have the same hash value.
```

To avoid this kind of confusing behavior, `__hash__` should either be removed or made to match the contract.  Seems fine to just remove it.

**Comment by vishwakftw** [comment]
I think it should be removed. Give me time until the weekend, and I should be able to send in a PR to fix this issue (hopefully).

cc: @weiyangfb

**Comment by fritzo** [comment]
Gosh I find `Tensor.__hash__()` quite convenient. While `Tensor.__hash__()` is incorrect for dicts with arbitrary keys (in which case `id(tensor)` can be used instead), it is perfectly correct for dicts with only tensors as keys (proof: `__eq__` is only called on hash collision, but `Tensor.__hash__()` hashes by id, hence there will never be a hash collision).

It seems reasonable to me to support at least the tensor-only key scenario.

**Comment by sdaulton** [comment]
What's the status of this issue? I was surprised to find that nn.Modules are hashable. I discovered this because if you write a custom __eq__ method for a class that inherits from nn.Module, you also need to write a custom __hash__ method to avoid breaking code in places where Modules are hashed (e.g. https://github.com/vishwakftw/pytorch/blob/master/torch/nn/modules/module.py#L1067).  Is there a reason that https://github.com/pytorch/pytorch/pull/10625 was never merged in?

**Comment by mruberry** [comment]
> What's the status of this issue? I was surprised to find that nn.Modules are hashable. I discovered this because if you write a custom **eq** method for a class that inherits from nn.Module, you also need to write a custom **hash** method to avoid breaking code in places where Modules are hashed (e.g. https://github.com/vishwakftw/pytorch/blob/master/torch/nn/modules/module.py#L1067). Is there a reason that #10625 was never merged in?

I don't think anyone is currently working on this.

It looks like #10625 was dropped because of concerns about the impact of the change on the PyTorch community. Changing a tensor's hashing behavior requires careful consideration, and may have surprising effects on PyTorch libraries and programs.

**Comment by dodobyte** [comment]
unlike python, pytorch's __hash__ function seems to give different hash values for the same underlying data, so it's not really useful for @lru_cache unless you're calling the function with the same exact tensor.
